tests/MongoDB.Driver.Core.Tests/Core/Authentication/MongoAWSAuthenticatorTests.cs

```diff
@@ -0,0 +1,188 @@
+﻿/* Copyright 2020–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Net;
+using System.Threading;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Servers;
+using MongoDB.Driver.Core.Helpers;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.WireProtocol.Messages;
+using Xunit;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Authentication;
+using Moq;
+using MongoDB.Bson.IO;
+
+namespace MongoDB.Driver.Core.Tests.Core.Authentication
+{
+    public class MongoAWSAuthenticatorTests
+    {
+        // private constants
+        private const int ClientNonceLength = 32;
+
+        // private static
+        private static readonly IRandomByteGenerator __randomByteGenerator = new DefaultRandomByteGenerator();
+
+        [Theory]
+        [ParameterAttributeData]
+        public void AuthenticateShouldWorkAsExpected(
+            [Values(false, true)] bool async)
+        {
+            var dateTime = DateTime.UtcNow;
+
+            var serverId = new ServerId(new ClusterId(), new DnsEndPoint("localhost", 27017));
+            var connectionDescription = new ConnectionDescription(
+                new ConnectionId(serverId),
+                new IsMasterResult(new BsonDocument("ok", 1).Add("ismaster", 1)),
+                new BuildInfoResult(new BsonDocument("version", "4.0.0")));
+
+            var clientNonce = __randomByteGenerator.Generate(ClientNonceLength);
+            var serverNonce = Combine(clientNonce, __randomByteGenerator.Generate(ClientNonceLength));
+            var host = "sts.amazonaws.com";
+            var credential = new UsernamePasswordCredential("$external", "permanentuser", "FAKEFAKEFAKEFAKEFAKEfakefakefakefakefake");
+            
+            AwsSignatureVersion4.SignRequest(
+                dateTime,
+                credential.Username,
+                credential.GetInsecurePassword(),
+                null,
+                serverNonce,
+                host,
+                out var authHeader,
+                out var timestamp);
+
+            var dateTimeProviderMock = new Mock<IClock>();
```

Comment:
nit: name variable "mockClock"

I suppose putting "mock" either at the front or the back is grammatically correct depending on whether you consider "mock" to be an adjective or a noun.

But our practice is to use "mockXyz" instead of "xyzMock".